diff --git a/configure.ac b/configure.ac
index f15d9ff..7de1017 100644
--- a/configure.ac
+++ b/configure.ac
@@ -64,7 +64,7 @@ m4_ifdef([PKG_PROG_PKG_CONFIG],[],[
 PKG_PROG_PKG_CONFIG()
 
 dnl LIB_DIR - library basename
-LIB_DIR=`echo $libdir | $SED 's%.*/%%'`
+LIB_DIR=`echo $libdir | $SED 's%[[^/]*/[^/]]*/%%'`
 AC_SUBST([LIB_DIR])
 
 dnl Cache LDFLAGS so we can add EXTRA_LIB_PATH and restore it later

