diff --git a/src/gbm/Makefile.am b/src/gbm/Makefile.am
index e22c55c..556eb03 100644
--- a/src/gbm/Makefile.am
+++ b/src/gbm/Makefile.am
@@ -18,7 +18,7 @@ libgbm_la_SOURCES = \
 	main/gbm.c \
 	main/backend.c \
 	main/common.c
-libgbm_la_LDFLAGS = -version-info 1:0
+libgbm_la_LDFLAGS = -version-info 1:0 -Wl,-z,defs
 libgbm_la_LIBADD = $(LIBUDEV_LIBS) $(LIBKMS_LIBS) $(DLOPEN_LIBS)
 
 if HAVE_EGL_PLATFORM_WAYLAND
