diff -up mesa-20080612/src/mesa/drivers/osmesa/Makefile.osmesa mesa-20080612/src/mesa/drivers/osmesa/Makefile
--- mesa-20080612/src/mesa/drivers/osmesa/Makefile.osmesa	2008-06-12 12:17:27.000000000 +1000
+++ mesa-20080612/src/mesa/drivers/osmesa/Makefile	2008-06-12 13:14:24.000000000 +1000
@@ -47,7 +47,7 @@ osmesa8: $(TOP)/lib/$(OSMESA_LIB_NAME)
 $(TOP)/lib/$(OSMESA_LIB_NAME): $(OBJECTS)
 	$(TOP)/bin/mklib -o $(OSMESA_LIB) \
 		-linker "$(CC)" \
-		-major $(MESA_MAJOR) -minor $(MESA_MINOR) -patch $(MESA_TINY) \
+		-major 6 -minor 5 -patch 3 \
 		-install $(TOP)/$(LIB_DIR) \
 		$(MKLIB_OPTIONS) \
 		$(OSMESA_LIB_DEPS) $(OBJECTS)
@@ -60,7 +60,7 @@ $(TOP)/lib/$(OSMESA_LIB_NAME): $(OBJECTS
 osmesa16: $(OBJECTS) $(CORE_MESA)
 	$(TOP)/bin/mklib -o $(OSMESA_LIB) \
 		-linker "$(CC)" \
-		-major $(MESA_MAJOR) -minor $(MESA_MINOR) -patch $(MESA_TINY) \
+		-major 6 -minor 5 -patch 3 \
 		-install $(TOP)/$(LIB_DIR) \
 		$(MKLIB_OPTIONS) \
 		$(OSMESA_LIB_DEPS) $(OBJECTS) $(CORE_MESA)
