---
 src/mesa/drivers/osmesa/Makefile |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/src/mesa/drivers/osmesa/Makefile.am
+++ b/src/mesa/drivers/osmesa/Makefile.am
@@ -35,7 +35,7 @@
 
 lib@OSMESA_LIB@_la_SOURCES = osmesa.c
 
-lib@OSMESA_LIB@_la_LDFLAGS = -module -version-number @OSMESA_VERSION@ -shared -no-undefined
+lib@OSMESA_LIB@_la_LDFLAGS = -module -version-number 6.5.3 -shared -no-undefined
 lib@OSMESA_LIB@_la_LIBADD = \
 	$(top_builddir)/src/mesa/libmesa.la \
 	$(top_builddir)/src/mapi/glapi/libglapi.la
