Index: mesa/src/gallium/auxiliary/rtasm/rtasm_execmem.c
===================================================================
--- mesa.orig/src/gallium/auxiliary/rtasm/rtasm_execmem.c	2012-01-18 16:04:58.230651581 +0200
+++ mesa/src/gallium/auxiliary/rtasm/rtasm_execmem.c	2012-01-18 16:06:09.872798989 +0200
@@ -37,7 +37,7 @@
 
 #include "rtasm_execmem.h"
 
-#if defined(PIPE_OS_BSD)
+#ifndef MAP_ANONYMOUS
 #define MAP_ANONYMOUS MAP_ANON
 #endif
 
Index: mesa/src/gallium/include/pipe/p_config.h
===================================================================
--- mesa.orig/src/gallium/include/pipe/p_config.h	2012-01-18 16:04:58.238651821 +0200
+++ mesa/src/gallium/include/pipe/p_config.h	2012-01-18 16:06:09.872798989 +0200
@@ -171,7 +171,7 @@
 #define PIPE_OS_ANDROID
 #endif
 
-#if defined(__FreeBSD__)
+#if defined(__FreeBSD__) || defined(__FreeBSD_kernel__)
 #define PIPE_OS_FREEBSD
 #define PIPE_OS_BSD
 #define PIPE_OS_UNIX
