Index: mesa/src/mapi/glapi/glapi_x86.S
===================================================================
--- mesa.orig/src/mapi/glapi/glapi_x86.S	2012-06-01 13:02:45.863167925 -0400
+++ mesa/src/mapi/glapi/glapi_x86.S	2012-06-01 13:02:46.203167935 -0400
@@ -147,7 +147,6 @@
 
 		ALIGNTEXT16
 		GLOBL GLNAME(gl_dispatch_functions_start)
-		HIDDEN(GLNAME(gl_dispatch_functions_start))
 GLNAME(gl_dispatch_functions_start):
 
 	GL_STUB(NewList, 0, NewList@8)
