--- a/src/mapi/glapi/gen/gl_x86_asm.py
+++ b/src/mapi/glapi/gen/gl_x86_asm.py
@@ -176,7 +176,6 @@ class PrintGenericStubs(gl_XML.gl_print_
         print ''
         print '\t\tALIGNTEXT16'
         print '\t\tGLOBL GLNAME(gl_dispatch_functions_start)'
-        print '\t\tHIDDEN(GLNAME(gl_dispatch_functions_start))'
         print 'GLNAME(gl_dispatch_functions_start):'
         print ''
         return
