From 194d039f1efee0e666c2d1b1116fd5adea8cb942 Mon Sep 17 00:00:00 2001
From: Kristian Høgsberg<krh@redhat.com>
Date: Wed, 21  Jan  2009  16:47:01  +0000
Subject: [intel] Remove remaining references to intel_wait_flips().

Oops.
---
diff --git a/src/mesa/drivers/dri/i915/intel_tris.c b/src/mesa/drivers/dri/i915/intel_tris.c
index c4708dc..e809965 100644
--- a/src/mesa/drivers/dri/i915/intel_tris.c
+++ b/src/mesa/drivers/dri/i915/intel_tris.c
@@ -92,7 +92,6 @@ static void intel_start_inline(struct intel_context *intel, uint32_t prim)
    BATCH_LOCALS;
    uint32_t batch_flags = LOOP_CLIPRECTS;
 
-   intel_wait_flips(intel);
    intel->vtbl.emit_state(intel);
 
    intel->no_batch_wrap = GL_TRUE;
@@ -214,8 +213,7 @@ void intel_flush_prim(struct intel_context *intel)
       return;
 
    /* Clear the current prims out of the context state so that a batch flush
-    * flush triggered by wait_flips or emit_state doesn't loop back to
-    * flush_prim again.
+    * flush triggered by emit_state doesn't loop back to flush_prim again.
     */
    vb_bo = intel->prim.vb_bo;
    dri_bo_reference(vb_bo);
@@ -227,8 +225,6 @@ void intel_flush_prim(struct intel_context *intel)
       intel->prim.start_offset = ALIGN(intel->prim.start_offset, 128);
    intel->prim.flush = NULL;
 
-   intel_wait_flips(intel);
-
    intel->vtbl.emit_state(intel);
 
    aper_array[0] = intel->batch->buf;
--
cgit v0.8.1-24-ge5fb
