diff --git a/src/mesa/drivers/dri/radeon/radeon_screen.c b/src/mesa/drivers/dri/radeon/radeon_screen.c
index ff65e6b..791daec 100644
--- a/src/mesa/drivers/dri/radeon/radeon_screen.c
+++ b/src/mesa/drivers/dri/radeon/radeon_screen.c
@@ -1159,6 +1159,8 @@ static void radeonDestroyContext(__DRIcontextPrivate * driContextPriv)
 static const __DRIconfig **
 radeonInitScreen(__DRIscreenPrivate *psp)
 {
+    const __DRIconfig **configs8, **configs16, **configs24, **configs32;
+
 #if !RADEON_COMMON
    static const char *driver_name = "Radeon";
    static const __DRIutilversion2 ddx_expected = { 4, 5, 0, 0 };
@@ -1206,11 +1208,15 @@ radeonInitScreen(__DRIscreenPrivate *psp)
    if (!radeonInitDriver(psp))
        return NULL;
 
-   return radeonFillInModes( psp,
-			     dri_priv->bpp,
-			     (dri_priv->bpp == 16) ? 16 : 24,
-			     (dri_priv->bpp == 16) ? 0  : 8,
-			     (dri_priv->backOffset != dri_priv->depthOffset) );
+    configs8  = radeonFillInModes(psp,  8,  8, 0, (dri_priv->backOffset != dri_priv->depthOffset));
+    configs16 = radeonFillInModes(psp, 16, 16, 0, (dri_priv->backOffset != dri_priv->depthOffset));
+    configs24 = radeonFillInModes(psp, 24, 24, 8, (dri_priv->backOffset != dri_priv->depthOffset));
+    configs32 = radeonFillInModes(psp, 32, 24, 8, (dri_priv->backOffset != dri_priv->depthOffset));
+
+    configs16 = driConcatConfigs(configs8, configs16);
+    configs24 = driConcatConfigs(configs16, configs24);
+
+   return driConcatConfigs(configs24, configs32);
 }
 
 
