Index: mesa/configure.ac
===================================================================
--- mesa.orig/configure.ac	2012-01-24 14:48:40.488101770 +0200
+++ mesa/configure.ac	2012-01-24 14:48:42.364157964 +0200
@@ -1770,7 +1770,7 @@
     if test "x$LLVM_CONFIG" != xno; then
 	LLVM_VERSION=`$LLVM_CONFIG --version | sed 's/svn.*//g'`
 	LLVM_CFLAGS=`$LLVM_CONFIG --cppflags|sed -e 's/-DNDEBUG\>//g' -e 's/-pedantic//g'`
-	LLVM_LIBS="`$LLVM_CONFIG --libs`"
+	LLVM_LIBS="-lLLVM-3.0"
 
 	LLVM_LDFLAGS=`$LLVM_CONFIG --ldflags`
 	DEFINES="$DEFINES -D__STDC_CONSTANT_MACROS"
