diff --git a/bin/mklib b/bin/mklib
index 2c7ed38..14bca63 100755
--- a/bin/mklib
+++ b/bin/mklib
@@ -147,6 +147,7 @@ ARCHOPT=""
 NOPREFIX=0
 EXPORTS=""
 ID=""
+WHOLE_ARCHIVE=""
 
 #
 # Parse arguments
@@ -191,6 +192,18 @@ do
 	-R*)
 	    DEPS="$DEPS $1"
 	    ;;
+	'-Wl,-whole-archive')
+	    shift
+	    while [ "x$1" != "x-Wl,-no-whole-archive" ] ; do
+		if [ "x$1" = "x" ] ; then
+		    echo "mklib: End of arguments reached when scanning for -Wl,-no-whole-archive"
+		    echo "mklib: Arguments to -no-whole-archive must be terminated by -Wl,-no-whole-archive"
+		    exit 1
+		fi
+		WHOLE_ARCHIVE="$WHOLE_ARCHIVE $1"
+		shift
+	    done
+	    ;;
 	-Wl*)
             DEPS="$DEPS $1"
             ;;
@@ -271,6 +284,10 @@ if [ $STATIC = 1 ]; then
     OBJECTS=$NEWOBJECTS
 fi
 
+if [ "x$WHOLE_ARCHIVE" != "x" ] ; then
+    OBJECTS="${OBJECTS} -Wl,-whole-archive ${WHOLE_ARCHIVE} -Wl,-no-whole-archive"
+fi
+
 
 #
 # Error checking
@@ -284,7 +301,6 @@ if [ "x${OBJECTS}" = "x" ] ; then
     exit 1
 fi
 
-
 #
 # Debugging info
 #
@@ -343,6 +359,7 @@ case $ARCH in
             fi
 
             rm -f ${LIBNAME}
+	    
             # make lib
             ${LINK} ${OPTS} ${LDFLAGS} -o ${LIBNAME} ${OBJECTS} ${DEPS}
             # finish up
diff --git a/configs/autoconf.in b/configs/autoconf.in
index 37b7ea5..b479f2b 100644
--- a/configs/autoconf.in
+++ b/configs/autoconf.in
@@ -121,6 +121,7 @@ DRI_DIRS = @DRI_DIRS@
 DRICORE_GLSL_LIBS = @DRICORE_GLSL_LIBS@
 DRICORE_LIBS = @DRICORE_LIBS@
 DRICORE_LIB_DEPS = @DRICORE_LIB_DEPS@
+GALLIUM_LIB_DEPS = @GALLIUM_LIB_DEPS@
 EGL_PLATFORMS = @EGL_PLATFORMS@
 EGL_CLIENT_APIS = @EGL_CLIENT_APIS@
 
diff --git a/configure.ac b/configure.ac
index 0ea264e..684f5b0 100644
--- a/configure.ac
+++ b/configure.ac
@@ -1036,6 +1036,41 @@ AC_SUBST([GLESv2_PC_LIB_PRIV])
 GLAPI_LIB_DEPS="-lpthread"
 AC_SUBST([GLAPI_LIB_DEPS])
 
+dnl Setup default DRI CFLAGS
+DRI_CFLAGS='$(CFLAGS)'
+DRI_CXXFLAGS='$(CXXFLAGS)'
+DRI_LIB_DEPS='$(TOP)/src/mesa/libmesa.a'
+MESA_MODULES='$(TOP)/src/mesa/libmesa.a'
+
+AC_ARG_ENABLE([shared-dricore],
+    [AS_HELP_STRING([--enable-shared-dricore],
+        [link DRI modules with shared core DRI routines @<:@default=disabled@:>@])],
+    [enable_dricore="$enableval"],
+    [enable_dricore=no])
+if test "$mesa_driver" = dri ; then
+   if test "$enable_dricore" = yes ; then
+      if test "$GCC$GXX" != yesyes ; then
+      	 AC_MSG_WARN([Shared dricore requires GCC-compatible rpath handling.  Disabling shared dricore])
+	 enable_dricore=no
+      else
+	 DRICORE_GLSL_LIBS='$(TOP)/$(LIB_DIR)/libglsl.so'
+	 DRICORE_LIBS='$(TOP)/$(LIB_DIR)/libdricore.so'
+	 DRICORE_LIB_DEPS='-L$(TOP)/$(LIB_DIR) -Wl,-R$(DRI_DRIVER_INSTALL_DIR) -lglsl'
+      	 DRI_LIB_DEPS='-L$(TOP)/$(LIB_DIR) -Wl,-R$(DRI_DRIVER_INSTALL_DIR) -lgallium -ldricore -lglsl'
+	 GALLIUM_LIB_DEPS='-lgallium'
+      	 DRI_CFLAGS='$(CFLAGS_NOVISIBILITY) -DUSE_DRICORE'
+      	 DRI_CXXFLAGS='$(CXXFLAGS_NOVISIBILITY) -DUSE_DRICORE'
+      	 MESA_MODULES='$(DRICORE_LIBS) $(DRICORE_GLSL_LIBS)'
+      fi
+   fi
+fi
+AC_SUBST([DRICORE_LIBS])
+AC_SUBST([DRICORE_GLSL_LIBS])
+AC_SUBST([DRICORE_LIB_DEPS])
+AC_SUBST([GALLIUM_LIB_DEPS])
+AC_SUBST([DRI_CXXFLAGS])
+AC_SUBST([DRI_CFLAGS])
+AC_SUBST([MESA_MODULES])
 
 dnl Setup default DRI CFLAGS
 DRI_CFLAGS='$(CFLAGS)'
diff --git a/src/gallium/Makefile.template b/src/gallium/Makefile.template
index 036c119..3f01d08 100644
--- a/src/gallium/Makefile.template
+++ b/src/gallium/Makefile.template
@@ -30,7 +30,7 @@ endif
 
 ##### TARGETS #####
 
-default: depend lib$(LIBNAME).a $(PROGS)
+default:: depend lib$(LIBNAME).a $(PROGS)
 
 lib$(LIBNAME).a: $(OBJECTS) $(EXTRA_OBJECTS) Makefile $(TOP)/src/gallium/Makefile.template
 	$(MKLIB) -o $(LIBNAME) -static $(OBJECTS) $(EXTRA_OBJECTS)
@@ -58,16 +58,16 @@ install:
 ##### RULES #####
 
 %.s: %.c
-	$(CC) -S $(INCLUDES) $(CFLAGS) $(LIBRARY_DEFINES) $< -o $@
+	$(CC) -S $(INCLUDES) $(DRI_CFLAGS) $(LIBRARY_DEFINES) $< -o $@
 
 %.o: %.c
-	$(CC) -c $(INCLUDES) $(CFLAGS) $(LIBRARY_DEFINES) $< -o $@
+	$(CC) -c $(INCLUDES) $(DRI_CFLAGS) $(LIBRARY_DEFINES) $< -o $@
 
 %.o: %.cpp
-	$(CXX) -c $(INCLUDES) $(CXXFLAGS) $(LIBRARY_DEFINES) $< -o $@
+	$(CXX) -c $(INCLUDES) $(DRI_CXXFLAGS) $(LIBRARY_DEFINES) $< -o $@
 
 %.o: %.S
-	$(CC) -c $(INCLUDES) $(CFLAGS) $(LIBRARY_DEFINES)  $< -o $@
+	$(CC) -c $(INCLUDES) $(DRI_CFLAGS) $(LIBRARY_DEFINES)  $< -o $@
 
 
 sinclude depend
diff --git a/src/gallium/auxiliary/Makefile b/src/gallium/auxiliary/Makefile
index 7dae7bc..9a08acb 100644
--- a/src/gallium/auxiliary/Makefile
+++ b/src/gallium/auxiliary/Makefile
@@ -213,6 +213,14 @@ endif
 
 include ../Makefile.template
 
+# Shared dricore library for classic DRI drivers
+$(TOP)/$(LIB_DIR)/libgallium.so: $(OBJECTS) $(EXTRA_OBJECTS)
+	@$(MKLIB) -o $@ -linker '$(CXX)' -ldflags '$(LDFLAGS)' \
+		-cplusplus -noprefix \
+		-install $(TOP)/$(LIB_DIR) -id $(DRI_DRIVER_INSTALL_DIR)/$@.dylib \
+		-Wl,-whole-archive $(TOP)/src/mesa/libgalliumcore.a -Wl,-no-whole-archive \
+		$(OBJECTS) $(EXTRA_OBJECTS)
+
 
 indices/u_indices_gen.c: indices/u_indices_gen.py
 	$(PYTHON2) $< > $@
@@ -228,4 +236,6 @@ util/u_format_table.c: util/u_format_table.py util/u_format_pack.py util/u_forma
 
 util/u_half.c: util/u_half.py
 	$(PYTHON2) util/u_half.py > $@
+
+default:: $(TOP)/$(LIB_DIR)/libgallium.so
 # DO NOT DELETE
diff --git a/src/gallium/targets/Makefile.dri b/src/gallium/targets/Makefile.dri
index 857ebfe..bb08ba0 100644
--- a/src/gallium/targets/Makefile.dri
+++ b/src/gallium/targets/Makefile.dri
@@ -9,7 +9,6 @@ LDFLAGS += -lstdc++
 endif
 
 MESA_MODULES = \
-	$(TOP)/src/mesa/libmesagallium.a \
 	$(GALLIUM_AUXILIARIES)
 
 COMMON_GALLIUM_SOURCES = \
@@ -78,8 +77,8 @@ $(LIBNAME): $(OBJECTS) $(MESA_MODULES) $(PIPE_DRIVERS) Makefile \
 	$(MKLIB) -o $@.tmp -noprefix -linker '$(CXX)' -ldflags '$(LDFLAGS)' \
 		$(OBJECTS) $(PIPE_DRIVERS) \
                 -Wl,--start-group $(MESA_MODULES) -Wl,--end-group \
-                 $(DRI_LIB_DEPS) $(DRIVER_EXTRAS)
-	$(CXX) $(CFLAGS) -o $@.test $(TOP)/src/mesa/drivers/dri/common/dri_test.o $@.tmp $(DRI_LIB_DEPS) $(LDFLAGS);
+                $(DRI_LIB_DEPS) $(GALLIUM_LIB_DEPS) $(DRIVER_EXTRAS)
+	$(CXX) $(CFLAGS) -o $@.test $(TOP)/src/mesa/drivers/dri/common/dri_test.o $@.tmp -lgallium $(DRI_LIB_DEPS) $(GALLIUM_LIB_DEPS) $(LDFLAGS);
 	@rm -f $@.test
 	mv -f $@.tmp $@
 
diff --git a/src/mesa/Makefile b/src/mesa/Makefile
index a903a26..c4ab43c 100644
--- a/src/mesa/Makefile
+++ b/src/mesa/Makefile
@@ -17,6 +17,9 @@ MESA_GALLIUM_OBJECTS := $(addprefix $(MESA_OBJ_DIR)/, $(MESA_GALLIUM_OBJECTS))
 
 DRICORE_OBJECTS := $(addprefix $(DRICORE_OBJ_DIR)/, $(MESA_OBJECTS))
 
+DRICORE_OBJECTS := $(addprefix $(DRICORE_OBJ_DIR)/, $(MESA_OBJECTS))
+GALLIUMCORE_OBJECTS := $(addprefix $(DRICORE_OBJ_DIR)/, $(filter-out $(MESA_OBJECTS), $(MESA_GALLIUM_OBJECTS)))
+
 # define preprocessor flags
 MESA_CPPFLAGS := $(API_DEFINES) $(DEFINES)
 
@@ -66,7 +69,7 @@ $(DRICORE_OBJ_DIR)/%.o: %.S
 
 # Default: build dependencies, then asm_subdirs, GLSL built-in lib,
 # then convenience libs (.a) and finally the device drivers:
-default: $(DEPENDS) asm_subdirs $(MESA_LIBS) $(DRICORE_LIBS) driver_subdirs
+default: $(DEPENDS) asm_subdirs $(MESA_LIBS) $(DRICORE_LIBS) libgalliumcore.a driver_subdirs
 
 main/api_exec_es1.c: main/APIspec.xml main/es_generator.py main/APIspecutil.py main/APIspec.py
 	$(PYTHON2) $(PYTHON_FLAGS) main/es_generator.py -S main/APIspec.xml -V GLES1.1 > $@
@@ -125,6 +128,10 @@ depend: $(ALL_SOURCES)
 	@$(MKDEP) $(MKDEP_OPTIONS) -p$(MESA_OBJ_DIR)/ $(MESA_CPPFLAGS) \
 		$(ALL_SOURCES) > /dev/null 2>/dev/null
 
+# Make archive of subset of core mesa object files for gallium
+libgalliumcore.a: $(GALLIUMCORE_OBJECTS)
+	@ $(MKLIB) -o galliumcore -static $(GALLIUMCORE_OBJECTS)
+
 ######################################################################
 # Installation rules
 
