--- a/configure.ac
+++ b/configure.ac
@@ -929,6 +929,7 @@
 AC_SUBST([GLESv2_PC_LIB_PRIV])
 
 DRI_LIB_DEPS="\$(top_builddir)/src/mesa/libdricore/libdricore${VERSION}.la"
+GALLIUM_DRI_LIB_DEPS="\$(TOP)/\$(LIB_DIR)/libdricore${VERSION}.so"
 
 AC_SUBST([HAVE_XF86VIDMODE])
 
--- a/src/gallium/targets/egl-static/Makefile
+++ b/src/gallium/targets/egl-static/Makefile
@@ -72,7 +72,7 @@
 egl_CPPFLAGS += -I$(TOP)/src/mesa $(API_DEFINES)
 # make st/mesa built-in when there is a single glapi provider
 ifeq ($(SHARED_GLAPI),1)
-egl_LIBS += $(TOP)/src/mesa/libmesagallium.a
+egl_LIBS += $(GALLIUM_DRI_LIB_DEPS)
 egl_SYS += -lm -lpthread $(DLOPEN_LIBS) -l$(GLAPI_LIB)
 else
 egl_CPPFLAGS += -D_EGL_EXTERNAL_GL=1
@@ -169,7 +169,7 @@
 
 # st_GL, built only when shared glapi is not enabled
 st_GL_CPPFLAGS := -I $(TOP)/src/mesa -I$(TOP)/src/gallium/include
-st_GL_LIBS := $(TOP)/src/mesa/libmesagallium.a $(GALLIUM_AUXILIARIES)
+st_GL_LIBS := $(GALLIUM_DRI_LIB_DEPS) $(GALLIUM_AUXILIARIES)
 st_GL_SYS := -lm -lpthread $(DLOPEN_LIBS)
 
 # LLVM
--- a/src/mesa/libdricore/Makefile.am
+++ b/src/mesa/libdricore/Makefile.am
@@ -45,6 +45,9 @@
 	$(LIBGLSL_CXX_FILES) \
 	$(BUILTIN_COMPILER_GENERATED_CXX_FILES) \
 	$(top_builddir)/src/glsl/builtin_function.cpp
+if HAVE_GALLIUM
+libdricore@VERSION@_la_SOURCES += $(STATETRACKER_FILES) $(SRCDIR)state_tracker/st_glsl_to_tgsi.cpp
+endif
 libdricore@VERSION@_la_LDFLAGS = -version-number 1:0
 libdricore@VERSION@_la_LIBADD = libdricore-asm.la
 
--- a/src/gallium/targets/libgl-xlib/Makefile
+++ b/src/gallium/targets/libgl-xlib/Makefile
@@ -46,7 +46,6 @@
 	$(TOP)/src/gallium/drivers/rbug/librbug.a \
 	$(TOP)/src/gallium/drivers/galahad/libgalahad.a \
 	$(TOP)/src/mapi/glapi/libglapi.a \
-	$(TOP)/src/mesa/libmesagallium.a \
 	$(GALLIUM_AUXILIARIES) \
 
 
--- a/src/mesa/Makefile.am
+++ b/src/mesa/Makefile.am
@@ -90,7 +90,7 @@
 check_LTLIBRARIES = libmesa.la
 endif
 if HAVE_GALLIUM
-noinst_LTLIBRARIES += libmesagallium.la
+#noinst_LTLIBRARIES += libmesagallium.la
 endif
 
 SRCDIR = $(top_srcdir)/src/mesa/
@@ -143,7 +143,7 @@
 # Provide compatibility with scripts for the old Mesa build system for
 # a while by putting a link to the library in the current directory.
 all-local: $(noinst_LTLIBRARIES)
-	ln -f .libs/libmesagallium.a .
+#	ln -f .libs/libmesagallium.a .
 endif
 
 CLEANFILES += libmesagallium.a
--- a/src/gallium/targets/Makefile.dri
+++ b/src/gallium/targets/Makefile.dri
@@ -11,7 +11,6 @@
 endif
 
 MESA_MODULES = \
-	$(TOP)/src/mesa/libmesagallium.a \
 	$(GALLIUM_AUXILIARIES)
 
 COMMON_GALLIUM_SOURCES = \
