diff --git a/src/gallium/auxiliary/util/u_math.h b/src/gallium/auxiliary/util/u_math.h
index 25bcfb5..d703bf0 100644
--- a/src/gallium/auxiliary/util/u_math.h
+++ b/src/gallium/auxiliary/util/u_math.h
@@ -40,6 +40,7 @@
 
 
 #include "pipe/p_compiler.h"
+#include "u_debug.h"
 
 
 #ifdef __cplusplus
@@ -807,7 +808,7 @@ util_memcpy_cpu_to_le32(void * restrict dest, const void * restrict src, size_t
 {
 #ifdef PIPE_ARCH_BIG_ENDIAN
    size_t i, e;
-   assert(n % 4 == 0);
+   debug_assert(n % 4 == 0);
 
    for (i = 0, e = n / 4; i < e; i++) {
       uint32_t * restrict d = (uint32_t* restrict)dest;
