Index: b/configure.ac
===================================================================
--- a/configure.ac
+++ b/configure.ac
@@ -1757,6 +1757,7 @@ gallium_check_st() {
 
 gallium_require_llvm() {
     if test "x$MESA_LLVM" = x0; then
+        case "$host" in *gnux32) return;; esac
         case "$host_cpu" in
         i*86|x86_64|amd64) AC_MSG_ERROR([LLVM is required to build $1 on x86 and x86_64]);;
         esac
